### ISSUE TITLE
haskell_cabal_* could use ghc instead of ghc for performance reasons

### ISSUE THREAD

**Issue by ulysses4ever** [tweag/rules_haskell#1051]
I noticed `cabal_wrapper.sh.tpl` uses `runghc`. [README](https://hackage.haskell.org/package/Cabal-3.0.0.0) of the Cabal library reads:

> Compiling 'Setup' rather than using runghc Setup is much faster and works on Windows. 

I thought maybe it makes sense to from `runghc` to `ghc` + calls to `./Setup`.

**Comment by mboes** [comment]
I'm not sure why that's the case though. The binary takes non-trivial time to link, whereas in `runghc` just has to typecheck a tiny source file and load the Cabal library (which is compiled) into ghci. Worth a try, but it would be good to measure the difference on a large package set (say all 200+ dependencies of Stack). It's also possible that the tradeoff is different when using a static GHC toolchain (see the `is_static` toolchain attribute).

What would most probably be a win though, is to precompile the default unmodified `Setup.hs`, then reuse the compiled binary whenever a package uses the default `Setup.hs`.

**Comment by ulysses4ever** [comment]
Makes sense, thank you for the reply. If you think this should be closed, feel free to do it. I might find the time to experiment with it in the Fall.

**Comment by mboes** [comment]
I think it's worth experimenting, so we can leave it open.
